File: src/main/java/com/sleepwalker/sleeplib/util/serialization/NetworkUtil.java
/*
 * Copyright (c)
 *
 * All rights reserved
 * Any copying and reproduction of the text, including partial and in any form,
 * without the written permission of the copyright holder is prohibited.
 * Any use of photos, audio, video and graphic materials, including partial,
 * without the written permission of the copyright holder is prohibited.
 *
 * © Салтыков Д.А., 2022
 */

package com.sleepwalker.sleeplib.util.serialization;

import net.minecraft.network.PacketBuffer;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.MathContext;
import java.util.*;
import java.util.function.BiConsumer;
import java.util.function.Function;

public final class NetworkUtil {

    public static void writeDecimal(@Nonnull PacketBuffer pf, @Nonnull BigDecimal bigDecimal){
        pf.writeByteArray(bigDecimal.unscaledValue().toByteArray());
        pf.writeVarInt(bigDecimal.scale());
    }

    @Nonnull
    public static BigDecimal readDecimal(@Nonnull PacketBuffer pf, @Nonnull MathContext context){
        BigInteger value = new BigInteger(pf.readByteArray());
        return new BigDecimal(value, pf.readVarInt(), context);
    }

    @Nonnull
    public static BigDecimal readDecimal(@Nonnull PacketBuffer pf){
        BigInteger value = new BigInteger(pf.readByteArray());
        return new BigDecimal(value, pf.readVarInt());
    }

    public static <K, V> void writeMap(
        @Nonnull PacketBuffer buffer, @Nonnull Map<K, V> map,
        @Nonnull BiConsumer<PacketBuffer, K> keyWriter, @Nonnull BiConsumer<PacketBuffer, V> valueWriter
    ) {
        buffer.writeVarInt(map.size());
        map.forEach((k, v) -> {
            keyWriter.accept(buffer, k);
            valueWriter.accept(buffer, v);
        });
    }

    @Nonnull
    public static <K, V> Map<K, V> readMap(
        @Nonnull PacketBuffer buffer,
        @Nonnull Function<PacketBuffer, K> keySupplier,
        @Nonnull Function<PacketBuffer, V> valueSupplier
    ) {

        int size = buffer.readVarInt();

        Map<K, V> map = new HashMap<>(size);

        for(int i = 0; i < size; i++){
            map.put(keySupplier.apply(buffer), valueSupplier.apply(buffer));
        }

        return map;
    }

    @Nonnull
    public static <K, V, M extends Map<K, V>> M readMap(
        @Nonnull PacketBuffer buffer,
        @Nonnull M map,
        @Nonnull Function<PacketBuffer, K> keySupplier,
        @Nonnull Function<PacketBuffer, V> valueSupplier
    ) {

        int size = buffer.readVarInt();

        for(int i = 0; i < size; i++){
            map.put(keySupplier.apply(buffer), valueSupplier.apply(buffer));
        }

        return map;
    }

    public static <V> void writeCollection(@Nonnull PacketBuffer buffer, @Nonnull Collection<V> collection,
                                           @Nonnull BiConsumer<PacketBuffer, V> valueConsumer) {
        buffer.writeVarInt(collection.size());
        collection.forEach(v -> valueConsumer.accept(buffer, v));
    }

    @Nonnull
    public static <V, C extends Collection<V>> C readCollection(
        @Nonnull PacketBuffer buffer,
        @Nonnull C collection,
        @Nonnull Function<PacketBuffer, V> valueSupplier
    ) {
        int size = buffer.readVarInt();
        for(int i = 0; i < size; i++){
            collection.add(valueSupplier.apply(buffer));
        }

        return collection;
    }

    @Nonnull
    public static <V> List<V> readArrayList(@Nonnull PacketBuffer buffer, @Nonnull Function<PacketBuffer, V> reader) {
        return readCollection(buffer, new ArrayList<>(), reader);
    }

    @Nonnull
    public static <V> Set<V> readHashSet(@Nonnull PacketBuffer buffer, @Nonnull Function<PacketBuffer, V> reader) {
        return readCollection(buffer, new HashSet<>(), reader);
    }

    public static void writeStringArray(@Nonnull PacketBuffer pf, @Nonnull String[] array){
        pf.writeVarInt(array.length);
        for(String str : array){
            pf.writeString(str);
        }
    }

    @Nonnull
    public static String[] readStringArray(@Nonnull PacketBuffer pf){
        int len = pf.readVarInt();
        String[] array = new String[len];
        for(int i = 0; i < len; i++){
            array[i] = pf.readString();
        }
        return array;
    }

    public static <T> void writeNullable(@Nonnull PacketBuffer buffer, @Nullable T value, @Nonnull BiConsumer<PacketBuffer, T> writer){
        buffer.writeBoolean(value != null);
        if(value != null){
            writer.accept(buffer, value);
        }
    }

    @Nullable
    public static <T> T readNullable(@Nonnull PacketBuffer buffer, @Nonnull Function<PacketBuffer, T> reader){
        if(buffer.readBoolean()){
            return reader.apply(buffer);
        }
        else {
            return null;
        }
    }
}

File: src/main/kotlin/com/sleepwalker/sleeplib/util/serialization/NetworkUtil.kt
package com.sleepwalker.sleeplib.util.serialization

import net.minecraft.network.PacketBuffer
import java.math.BigDecimal
import java.math.BigInteger
import java.math.MathContext
import javax.annotation.Nonnull

fun PacketBuffer.writeDecimal(bigDecimal: BigDecimal) {
    this.writeByteArray(bigDecimal.unscaledValue().toByteArray())
    this.writeVarInt(bigDecimal.scale())
}

fun PacketBuffer.readDecimal(context: MathContext): BigDecimal {
    val value = BigInteger(this.readByteArray())
    return BigDecimal(value, this.readVarInt(), context)
}

fun PacketBuffer.readDecimal(): BigDecimal {
    val value = BigInteger(this.readByteArray())
    return BigDecimal(value, this.readVarInt())
}

fun <K, V> PacketBuffer.writeMap(map: Map<K, V>, keyWriter: PacketBuffer.(K) -> Unit, valueWriter: PacketBuffer.(V) -> Unit) {
    this.writeVarInt(map.size)
    map.forEach { (k: K, v: V) ->
        keyWriter(this, k)
        valueWriter(this, v)
    }
}

fun <K, V> PacketBuffer.readMap(keyReader: PacketBuffer.() -> K, valueReader: PacketBuffer.() -> V): MutableMap<K, V> {
    val size = this.readVarInt()
    val map: MutableMap<K, V> = HashMap(size)
    for (i in 0 until size) {
        map[keyReader(this)] = valueReader(this)
    }
    return map
}

fun <K, V, M : MutableMap<K, V>> PacketBuffer.readMap(map: M, keyReader: PacketBuffer.() -> K, valueReader: PacketBuffer.() -> V): M {
    val size = this.readVarInt()
    for (i in 0 until size) {
        map.put(keyReader(this), valueReader(this))
    }
    return map
}

fun <V> PacketBuffer.writeCollection(collection: Collection<V>, valueWriter: PacketBuffer.(V) -> Unit) {
    this.writeVarInt(collection.size)
    collection.forEach { valueWriter(it) }
}

fun <V, C : MutableCollection<V>> PacketBuffer.readCollection(collection: C, valueReader: PacketBuffer.() -> V): C {
    val size = this.readVarInt()
    for (i in 0 until size) {
        collection.add(valueReader(this))
    }
    return collection
}

fun <V> PacketBuffer.readArrayList(valueReader: PacketBuffer.() -> V): MutableList<V> {
    return this.readCollection(ArrayList(), valueReader)
}

@Nonnull
fun <V> PacketBuffer.readHashSet(valueReader: PacketBuffer.() -> V): MutableSet<V> {
    return this.readCollection(HashSet(), valueReader)
}

fun PacketBuffer.writeStringArray(array: Array<String>) {
    this.writeVarInt(array.size)
    for (str in array) {
        this.writeString(str)
    }
}

fun PacketBuffer.readStringArray(): Array<String> {
    val len = this.readVarInt()
    val array = Array(len) { "" }
    for (i in 0 until len) {
        array[i] = this.readString()
    }
    return array
}

fun <T> PacketBuffer.writeNullable(value: T?, writer: PacketBuffer.(T) -> Unit) {
    this.writeBoolean(value != null)
    if (value != null) {
        writer(this, value)
    }
}

fun <T> PacketBuffer.readNullable(reader: PacketBuffer.() -> T): T? {
    return if (this.readBoolean()) reader(this) else null
}